RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc

```diff
@@ -197,348 +265,107 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
-                        layerClustersTimes,
-                        rhtools_.getPositionLayer(rhtools_.lastLayerEE()).z());
-  energyRegressionAndID(layerClusters, tfSession_, *resultTrackstersMerged);
-
-  printTrackstersDebug(*resultTrackstersMerged, "TrackstersMergeProducer");
-
-  auto trackstersMergedHandle = evt.put(std::move(resultTrackstersMerged));
-
-  // TICL Candidate creation
-  // We start from neutrals first
-
-  // Photons
-  for (unsigned i = 0; i < trackstersEM.size(); ++i) {
-    auto mergedIdx = indexInMergedCollEM[i];
-    usedTrackstersMerged[mergedIdx] = true;
-    const auto &t = trackstersEM[i];  //trackster
-    TICLCandidate tmpCandidate;
-    tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, mergedIdx));
-    tmpCandidate.setCharge(0);
-    tmpCandidate.setPdgId(22);
-    tmpCandidate.setRawEnergy(t.raw_energy());
-    math::XYZTLorentzVector p4(t.raw_energy() * t.barycenter().unit().x(),
-                               t.raw_energy() * t.barycenter().unit().y(),
-                               t.raw_energy() * t.barycenter().unit().z(),
-                               t.raw_energy());
-    tmpCandidate.setP4(p4);
-    resultCandidates->push_back(tmpCandidate);
-  }
+  // Linking
+  linkingAlgo_->linkTracksters(
+      track_h, trackTime, trackTimeErr, trackTimeQual, muons, trackstersclue3d_h, *resultCandidates);
 
-  // Neutral Hadrons
-  constexpr double mpion = 0.13957;
-  constexpr float mpion2 = mpion * mpion;
-  for (unsigned i = 0; i < trackstersHAD.size(); ++i) {
-    auto mergedIdx = indexInMergedCollHAD[i];
-    usedTrackstersMerged[mergedIdx] = true;
-    const auto &t = trackstersHAD[i];  //trackster
-    TICLCandidate tmpCandidate;
-    tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, mergedIdx));
-    tmpCandidate.setCharge(0);
-    tmpCandidate.setPdgId(130);
-    tmpCandidate.setRawEnergy(t.raw_energy());
-    float momentum = std::sqrt(t.raw_energy() * t.raw_energy() - mpion2);
-    math::XYZTLorentzVector p4(momentum * t.barycenter().unit().x(),
-                               momentum * t.barycenter().unit().y(),
-                               momentum * t.barycenter().unit().z(),
-                               t.raw_energy());
-    tmpCandidate.setP4(p4);
-    resultCandidates->push_back(tmpCandidate);
+  // Print debug info
+  if (debug_) {
+    LogDebug("TrackstersMergeProducer") << "Results from the linking step : " << std::endl
+                                        << "No. of Tracks : " << tracks.size()
+                                        << "  No. of Tracksters : " << (*trackstersclue3d_h).size() << std::endl
+                                        << "(neutral candidates have track id -1)" << std::endl;
   }
+  std::vector<TICLCandidate> &candidates = *resultCandidates;
+  for (const auto &cand : candidates) {
+    auto track_ptr = cand.trackPtr();
+    auto trackster_ptrs = cand.tracksters();
+    if (debug_) {
```

Comment:
this block could all be in `#ifdef EDM_ML_DEBUG`, as LogDebug is only enabled in that case.

Reply:
thanks, I have removed the runtime `debug_` flag completely and wrapped the actions associated to message logging in `#ifdef EDM_ML_DEBUG`